Classdef Improvements

ampreeT opened the pull request:
**Changes:**

- Converted spawn flags into individual functions for each bit
- Converted edict flags into individual functions for each bit
- Added `const` to most function parameters
- Added new `RGBA` methodmap
- Moved first-time gamedata initialization of classes to their own files
- Added `static` to internal implementation details (ex. handles and functions used by methodmaps)
- Renamed variables to be more consistent
- Replaced most instances of `CBaseEntity.IsValid()` with a check to `!= NULL_CBASEENTITY`

Alienmario commented on scripting/include/srccoop_api/classdef/core/CBaseEntity.inc:
```diff
@@ -441,52 +484,59 @@ methodmap CBaseEntity
+		int r = 0;
+		int g = 0;
+		int b = 0;
+		int a = 0;
```
can be on one line as well no need to init to 0

Alienmario commented on scripting/include/srccoop_api/classdef/core/CSoundEnt.inc:
```diff
@@ -1,11 +1,19 @@
 {
+	public static void InitClassdef(const GameData hGameConfig)
+	{
+		g_iCSoundSize = GetGamedataInt(hGameConfig, "CSoundSize");	// TODO: Change this to `hGameConfig.GetOffset`.
```
In other libraries I used sizes are defined as offsets like so `"sizeof(CSound)"`

ampreeT replied:
I like that a lot.

Changed the variable assignment to use `hGameConfig.GetOffset` and renamed gamedata variable to `sizeof(CSound)`.

Alienmario reviewed (commented):
Personally I'm not too sure making a function for every flag is a good thing. I like to know what I'm setting and this only makes cross referencing the sdk harder (even it does not do that for most flags). Then what about setting multiple flags / clearing / etc.

Same for renaming some [function names](https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/mp/src/game/shared/Sprite.h#L156).

Instead we could expose flags by properties rather than methods, as this would allow for direct bitwise operations on them (no more set+get).

I'd also like to keep spawnflags definitions for when not working with an entity directly. I think these can be at top of classdef files.

Otherwise it looks good, I like that the extra IsValid's are gone.

Lastly if we want to keep same naming structure everywhere, `classdef/core` should be `classdef/common`

ampreeT commented on scripting/include/srccoop_api/classdef/core/CBaseCombatCharacter.inc:
```diff
@@ -90,5 +105,15 @@ methodmap CBaseCombatCharacter < CBaseFlex
 		SetEntPropFloat(this.GetEntIndex(), Prop_Data, "m_flNextAttack", flNextAttack);
-	}	
+	}
+	public CBasePlayer AsPlayer()
```
Changed the functions to be prefixed with `To` instead of `As` to match the SDK.

ampreeT commented on scripting/include/srccoop_api/classdef/bms/CParamsManager.inc:
```diff
@@ -1,8 +1,16 @@
 #pragma newdecls required
 #pragma semicolon 1
 
+static int g_iOffset_CParamsManger_bIsMultiplayer = -1;
```
Fixed.

ampreeT commented on scripting/include/srccoop_api/classdef.inc:
```diff
@@ -9,55 +9,46 @@
-CPlayerInfoManager g_pPlayerInfoManager;
-
-ConVar ai_los_mode;
+stock bool GetBitFromInt(const int i, const int iBitIndex)
```
Fixed. Moved the 2 bit functions and 2 spawnflag functions to `utils.inc`.

ampreeT commented on scripting/include/srccoop_api/classdef.inc:
```diff
@@ -9,55 +9,46 @@
 #include <sdkhooks>
 #include <sdktools>
 #include <sourcescramble>
+#include <srccoop_api/color>
```
My plan was to eventually split up `utils.inc` and `typedef_game.inc` into multiple files under `srccoop_api` but I'm going to leave that for a new PR. As for `color.inc`, I think it would probably best to leave it for encapsulation incase we decide to expand further (HSL implementation, ect.).

Alienmario replied:
So was my plan too, already have some encapsulated classes under the util folder (what I mean)

ampreeT replied:
Ah I misunderstood. Moved `color.inc` into the `util` folder.

ampreeT commented on scripting/include/srccoop/blackmesa/idle_anims.inc:
```diff
@@ -27,6 +27,14 @@ enum struct IdleAnimEntry
+{
+	for (int i = 0; i < sizeof(g_pIdle); ++i)
+	{
+		g_pIdle[i].m_pModel = NULL_CBASEENTITY;
```
I tried the following but I got a error on the SourceMod version that we are using.

```
public IdleAnimEntry g_pIdle[MAXPLAYERS + 1] = { { NULL_CBASEENTITY, IdleAnimType_None }, ... };
```

```
error 047: array sizes do not match, or destination array is too small
    28 | public IdleAnimEntry g_pIdle[MAXPLAYERS + 1] = { { NULL_CBASEENTITY, IdleAnimType_None }, ... };
```

ampreeT commented on scripting/include/srccoop/blackmesa/idle_anims.inc:
```diff
@@ -8,7 +8,7 @@
 enum IdleAnimType
 {
-	None = 0,
+	None = 0,	// TODO: Will this having a naming conflict with other enums? If so, rename.
```
Fixed by prefixing the enum name to each of the enum's variants.

ampreeT commented:
> Personally I'm not too sure making a function for every flag is a good thing. I like to know what I'm setting and this only makes cross referencing the sdk harder (even it does not do that for most flags). Then what about setting multiple flags / clearing / etc.
> 
> Same for renaming some [function names](https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/mp/src/game/shared/Sprite.h#L156).
> 
> Instead we could expose flags by properties rather than methods, as this would allow for direct bitwise operations on them (no more set+get).
> 
> I'd also like to keep spawnflags definitions for when not working with an entity directly. I think these can be at top of classdef files.
> 
> Otherwise it looks good, I like that the extra IsValid's are gone.
> 
> Lastly if we want to keep same naming structure everywhere, `classdef/core` should be `classdef/common`

I wasn't a fan of the original way I did it with making a function for each spawn flag bit. I ended up going with properties + enum definitions for the variants.

I also fixed the incorrectly named function that you linked to match the SDK name and renamed the `core` folder to `common`.

Alienmario reviewed (approved):
Looks much better, thanks!
